sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala

```diff
@@ -176,4 +186,13 @@ class DatasetAggregatorSuite extends QueryTest with SharedSQLContext {
         typed.avg(_._2), typed.count(_._2), typed.sum(_._2), typed.sumLong(_._2)),
       ("a", 2.0, 2L, 4.0, 4L), ("b", 3.0, 1L, 3.0, 3L))
   }
+
+  test("SPARK-12555 - result should not be corrupted after input columns are reordered") {
+    val ds = sql("SELECT 1279869254 AS a, 'Some String' AS b").as[AggData]
```

Comment:
the scenario in the JIRA is
```
case class people(age: Int, name: String)
sqlContext.sql("SELECT 'Tim Preece' AS name, 1279869254 AS age").as[people]
```
I think input column reorder is important to trigger this bug, could you try your test case at branch 1.6 and see if it reproduce? thanks